regression-tests/modules/dlso-gsql/dlsogsql.cc

```diff
@@ -0,0 +1,661 @@
+    const struct resource_record* record = &(records[i]);
+
+    DNSResourceRecord rr;
+    rr.qtype = record->qtype;
+    string qname_ = string(record->qname, record->qname_len);
+    rr.qname = DNSName(qname_);
+    rr.qclass = QClass::IN;
+    rr.content = string(record->content, record->content_len);
+    rr.ttl = record->ttl;
+    rr.auth = record->auth;
+    rr.scopeMask = record->scope_mask;
+    rr.domain_id = record->domain_id;
+
+    rrset.push_back(rr);
+  }
+
+  auto qtype_ = QType(qtype);
+  auto qname_ = DNSName(string(qname, qlen));
+
+  return handle->module->replaceRRSet(domain_id, qname_, qtype_, rrset);
+}
+
+bool add_record_ent(void* ptr, uint32_t domain_id, bool auth, uint8_t qlen, const char* qname_)
+{
+  auto* handle = static_cast<struct dlso_gsql*>(ptr);
+
+  auto qname = DNSName(string(qname_, qlen));
+  map<DNSName, bool> nonterm;
+  nonterm.insert({qname, auth});
+
+  return handle->module->feedEnts(domain_id, nonterm);
+}
+
+bool add_record_ent_nsec3(void* ptr, uint32_t domain_id, uint8_t domain_len, const char* domain_, bool narrow, bool auth, uint8_t qlen, const char* qname_, const struct nsec3_param* ns3)
+{
+  auto* handle = static_cast<struct dlso_gsql*>(ptr);
+
+  auto qname = DNSName(string(qname_, qlen));
+  map<DNSName, bool> nonterm;
+  nonterm.insert({qname, auth});
+
+  DNSName domain;
+  if (domain_len > 0) {
+    domain = DNSName(string(domain_, domain_len));
+  }
+
+  NSEC3PARAMRecordContent ns3prc;
+  ns3prc.d_algorithm = ns3->alg;
+  ns3prc.d_flags = ns3->flags;
+  ns3prc.d_iterations = ns3->iterations;
+  ns3prc.d_salt = string(ns3->salt, ns3->salt_len);
+
+  return handle->module->feedEnts3(domain_id, domain, nonterm, ns3prc, narrow);
+}
+
+std::mutex g_configuration_mutex;
+
+extern "C" bool pdns_dlso_register(uint32_t abi_version, struct lib_so_api** ptr, bool dnssec, const char* args)
+{
+  auto* gsql = static_cast<struct dlso_gsql*>(malloc(sizeof(struct dlso_gsql)));
+  if (gsql == nullptr) {
+    return false;
+  }
+  auto* api = static_cast<struct lib_so_api*>(malloc(sizeof(struct lib_so_api)));
+  if (api == nullptr) {
+    free(gsql);
+    return false;
+  }
+
+  memset(api, 0, sizeof(*api));
+  *ptr = api;
+
+  // Configuration and its underlying std::map does not allow two thread
+  // to write to it concurrently. When two threads concurrently
+  // declareArguments() the map could do double-free. This is bad and this
+  // shouldn't happen.
+  // Powerdns normally does declare argument and parse configuration
+  // in its very beginning in a single-thread. This is not ArgvMap
+  // responsability to protect from current access. This test backend
+  // being a wrapper Our test backend here may be executed multiple times
+  // by different threads (signing thread, packet receiver)
+  // For this reason, I chose to have a simple mutex, to not allow two
+  // threads to register simultinaeously.
+  std::lock_guard<std::mutex> guard(g_configuration_mutex);
```

Comment:
I would suggest wrapping the data you want to protect in a LockGuarded object, see `lock.hh`.